### ISSUE TITLE
OIDC signout should return to CallbackPath then locally redirect to AuthProperties.RedirectUri

### ISSUE THREAD

**Issue by Tratcher** [aspnet/Security#743]
Feedback from template reviews:

Default PostLogoutRedirectUri to CallbackPath. Listen on that path for signout callbacks and then apply the local redirect specified in auth properties. This prevents needing to register a separate redirect uri address with the provider.

If we do this do we still need PostLogoutRedirectUri? How do we reconcile it with AuthProperties.RedirectUri if both are set?

@vibronet @phenning @rustd

**Comment by PinpointTownes** [comment]
Interesting approach, I like it. That said, it might be tricky to implement, since there's no logout-specific parameter: http://openid.net/specs/openid-connect-session-1_0.html#RPLogout.

A way to implement that would be using the `state` parameter to store the callback type.

**Comment by Tratcher** [comment]
When looking into this I found that:
- AAD does not restrict the value of post_logout_redirect_uri like it should
- AAD does not flow state to the post_logout_redirect_uri like it should
- Google doesn't implement end_session_endpoint at all

Putting this in backlog until AAD gets this sorted out on their end.

**Comment by brockallen** [comment]
> AAD does not flow state to the post_logout_redirect_uri like it should

I was just testing with AAD today (via OIDC) and it was passing back the state param on post signout redirects.

**Comment by Tratcher** [comment]
@brockallen Hmm, I just tested it today as well.

Here's what my initial signout redirect looks like:
```
Location: https://login.microsoftonline.com/cdc690f9-b6b8-4023-813a-bae7143d1f87/oauth2/logout?post_logout_redirect_uri=http%3A%2F%2Flocalhost%3A42023%2Fsignin-oidc&state=CfDJ8AaEPKTYgPVMjEyowKKgeJKyKJt3ENUSWXDuMtRWlpHb716h5OpgKR22nGFad0K5mdPpAJHZVCOI6sxfx0GYP3TPBgWXseU2C0ffXVFyMusoLj9UJ69uhyJ9ypTT7M15WuBZ_6NEPIKNIUTO-mYQ_zyA6wy27iGN0cQMwuH-PIEB
```
But all I get back is:
```
Location: http://localhost:42023/signin-oidc
```

How about you?

**Comment by brockallen** [comment]
```
https://login.windows.net/4ca9cb4c-5e5f-4be9-b700-c532992a3705/oauth2/logout?post_logout_redirect_uri=https%3a%2f%2flocalhost%3a44333%2fcore%2flogout%2f%3fstate%3dfc795bfc9a1b58f2ce6a4b4e43920c2c
```

**Comment by PinpointTownes** [comment]
@brockallen there's a huge difference between your logout request and the one sent by @Tratcher.

In your case, the state is appended to the `post_logout_redirect_uri` URL while @Tratcher used the standard `state` parameter to flow it.

**Comment by brockallen** [comment]
Oh, yes good point. I had the state param on my mind, but hadn't noticed the difference.

**Comment by Tratcher** [comment]
Followup: AAD does restrict post_logout_redirect_uri to the same host, but it doesn't mater what the path is. Postponing this feature until further notice.

**Comment by dstrockis** [comment]
Hey @Tratcher, I got tripped up by this today.  I noticed that in the current release `AuthProperties.RedirectUri` is being used to specify the value of the `post_logout_redirect_uri` in the sign out request.  `OpenIDConnectOptions.PostLogoutRedirectUri` is totally unused.  

In fact, if I *do* specify `OpenIDConnectOptions.PostLogoutRedirectUri` and *do not* specify `AuthProperties.RedirectUri`, the middleware does not send any `post_logout_redirect_uri` value.  And in the `RedirectToIdentityProviderForSignOut` event, `Options.PostLogoutRedirectUri` was null.  It seems like the middleware was dropping that setting entirely.

I'm not recommending any particular approach, just noticing that the current behavior is a little confusing.

And re: AAD behavior.  Yes, the above is correct.  But we also have a new "v2.0" protocol implementation which has more spec-compliant behavior.  For instance, it *does* perform strict redirect_uri matching and *does* return the state in the sign-out request.
